### ISSUE TITLE
Integration Tracking (WC Subscriptions): Provide extensions a way to signal creating an account is required for the current cart

### ISSUE THREAD

**Issue by haszari** [woocommerce/woocommerce-gutenberg-products-block#3253]
Extensions like Subscriptions require an account to create a subscription.
There's nothing to force the checkout to create an account if this value is set
![image](https://user-images.githubusercontent.com/6165348/95477745-94180c80-0980-11eb-91c1-a293ec3c26ea.png)

So a shopper can try to checkout directly and he will see this error
<img width="1252" alt="Screen Shot 2020-10-08 at 1 59 56 PM" src="https://user-images.githubusercontent.com/4167300/95403493-ddf7f700-096e-11eb-8cf0-9c0286616b2e.png">

We need a way to signal that creating an account is required regardless of that option and on a cart basis (not a global setting).

This can be done PHP side or JS side, but I'd prefer PHP side.

**Comment by senadir** [comment]
Subscriptions require an account, it seems our "Create an account?" is not working well with it, there's a weird sequence:

- On a checkout with incognito, check "Create an account?" and checkout.
- You will see the error about customer_id.
- Clicking the checkout button again will give you "X-WC-Store-API-Nonce is invalid."
- Refresh the page, and "Create an account?" is not there anymore.
- Checking out will pass now and an account will be created with the subscription.

**Comment by senadir** [comment]
We should refresh the order object once we create an account, so after this line
https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/add/order-processed-action/src/StoreApi/Routes/Checkout.php#L172

**Comment by senadir** [comment]
So I'm going to rename this function and update its content, I hope that's okay with you Rua.
Subscriptions don't create an account, it expects an account to exist in order to continue, there was a bug in the create account code that caused this issue, I fixed it in #3260 
However, if a customer chooses not to create an account, they will see that error.
Extensions (like Subscriptions) need a way to signal to the checkout order that creating accounts is required 
This means, overriding this setting value
![image](https://user-images.githubusercontent.com/6165348/95477235-089e7b80-0980-11eb-9bf3-e463208f20c5.png)

So I'm going to update the issue accordingly.

**Comment by haszari** [comment]
Update: looks like this is changing (how it works) in Woo Subscriptions, based on changes in Woo core 4.6.2: https://github.com/woocommerce/woocommerce-subscriptions/issues/3870

Specifically, Subs is looking at using  [`woocommerce_checkout_registration_required` or `woocommerce_checkout_registration_enabled` hooks](https://woocommerce.github.io/code-reference/files/woocommerce-includes-class-wc-checkout.html#source-view.184) to signal to Woo / checkout that an account is required. 

I think that would address this too, as we currently consult [`is_registration_required()`](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/d1c0796361b4f6e94211336b59ac9fa44655bc1d/src/Domain/Services/CreateAccount.php#L154).

cc @senadir  @james-allan since you're working on this :)

**Comment by senadir** [comment]
I asked James before about some pointer for when 4.7 was introduced, this is the (old?) way of doing things
p1604618593376400-slack-CJTS771PB

**Comment by senadir** [comment]
By having account creation required, we would need to remove the checkbox in our checkbox, so we need to separate between `CHECKOUT_ALLOWS_GUEST` and `CHECKOUT_ALLOWS_SIGNUP`.
The first means accounts are required.

**Comment by haszari** [comment]
> By having account creation required, we would need to remove the checkbox in our checkbox, so we need to separate between CHECKOUT_ALLOWS_GUEST and CHECKOUT_ALLOWS_SIGNUP.
The first means accounts are required.

We already support this - if accounts are required, the checkbox is hidden.

- [We only show the checkbox if `CHECKOUT_ALLOWS_GUEST`](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/8f335014488f2d797087c9f7a4a0b78503f27597/assets/js/blocks/cart-checkout/checkout/form/contact-fields-step.js#L30)
- [In back end, if registration is required, we create an account](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/d1c0796361b4f6e94211336b59ac9fa44655bc1d/src/Domain/Services/CreateAccount.php#L154)

**Comment by senadir** [comment]
To summarize the above slack discussion:

Adding and making the account fields required:
Hook - https://github.com/woocommerce/woocommerce-subscriptions/blob/3.0.9/includes/class-wc-subscriptions-checkout.php#L30
Callback - https://github.com/woocommerce/woocommerce-subscriptions/blob/3.0.9/includes/class-wc-subscriptions-checkout.php#L437
Hook - https://github.com/woocommerce/woocommerce-subscriptions/blob/3.0.9/includes/class-wc-subscriptions-checkout.php#L27
Callback - https://github.com/woocommerce/woocommerce-subscriptions/blob/3.0.9/includes/class-wc-subscriptions-checkout.php#L418
There are some other guest checkout disabling done via WC core JS script args:
Hook - https://github.com/woocommerce/woocommerce-subscriptions/blob/3.0.9/includes/class-wc-subscriptions-checkout.php#L57
Callback - https://github.com/woocommerce/woocommerce-subscriptions/blob/3.0.9/includes/class-wc-subscriptions-checkout.php#L482

This is the old way of doing this, in 3883-gh-woocommerce/woocommerce-subscriptions/ WCS is going to support those filters you mentioned Rua, making this case supported by default (and therefore closing this issue).

We're going to wait on the niche option of enabling account creation for Subs only, I guess it should be supported by default looking at the code, but if it's not, it would be out of scope here.

**Comment by haszari** [comment]
Thanks for linking in that detailed info @senadir . Marking this PR as blocked on https://github.com/woocommerce/woocommerce-subscriptions/issues/3870 and PR https://github.com/woocommerce/woocommerce-subscriptions/pull/3883. When that merges we can revisit this and close or address any follow-up needed.

**Comment by senadir** [comment]
This is merged as part of 3.1 release, so we should be okay closing.
